### PR TITLE
Implementation of Advanced BIT* (ABIT*)

### PR DESCRIPTION
**Pull request by marlinstrub**
Hi

My name is Marlin and I'm currently a PhD student in Jon @gammell 's lab. We've been building on BIT*'s core ideas and one of the algorithms that we developed is Advanced BIT* (ABIT*). In order to avoid code duplication, I've modified BIT*'s implementation such that protected methods can turn it into ABIT*. This allows ABIT*'s implementation to simply inherit from BIT* and configure it appropriately. 

ABIT* is to appear in ICRA 2020 and is already available [on arxiv](https://arxiv.org/abs/2002.06589). A short video explaining the algorithm is available [on youtube](https://youtu.be/VFdihv8Lq2A).

Please let me know if there's anything I can do to make reviewing this PR easier for you. Thank you for maintaining OMPL.

### CODE REVIEW

File: src/ompl/geometric/planners/bitstar/BITstar.h [hunk 1]
```diff
@@ -375,118 +421,135 @@ namespace ompl
...
             std::shared_ptr<ImplicitGraph> graphPtr_{nullptr};
 
             /** \brief The queue of vertices to expand and edges to process ordered on "f-value", i.e., estimated
              * solution cost. Remaining vertex queue "size" and edge queue size are accessible via
              * vertexQueueSizeProgressProperty and edgeQueueSizeProgressProperty, respectively. */
             std::shared_ptr<SearchQueue> queuePtr_{nullptr};
 
-            /** \brief The goal vertex of the current best solution */
+            /** \brief The inflation factor for the initial search. */
+            double initialInflationFactor_{1.0};
```
Comment (mamoll) [inline src/ompl/geometric/planners/bitstar/BITstar.h:476]: I find the distinction between fooFactor_ and fooFactorParameter_ confusing. Can you think of a better naming convention to distinguish them? Also, I guess it's not possible to hide the ABIT* parameters and getter/setter methods from BIT*?

File: src/ompl/geometric/planners/bitstar/datastructures/CostHelper.h [hunk 2]
```diff
@@ -200,23 +212,21 @@ namespace ompl
...
-                return this->motionCost(edgePair.first->stateConst(), edgePair.second->stateConst());
+                return this->motionCost(edgePair.first->state(), edgePair.second->state());
             };
 
-            /** \brief Combine 3 costs */
-            inline ompl::base::Cost combineCosts(const ompl::base::Cost &a, const ompl::base::Cost &b,
-                                                 const ompl::base::Cost &c) const
+            /** \brief Combine multiple costs. */
+            template <typename... Costs>
+            inline ompl::base::Cost combineCosts(const ompl::base::Cost &cost, Costs... costs) const
```
Comment (mamoll) [inline src/ompl/geometric/planners/bitstar/datastructures/CostHelper.h:220]: Nice! Shouldn't that be `const Costs... &costs` though?

**Comment by mamoll** [comment]
@marlinstrub I'm trying to get a release of OMPL out soon that will be used for ROS 1 Noetic and the next ROS 2 release. Could you please address the last remaining issues? They seem relatively minor...

File: src/ompl/geometric/planners/bitstar/BITstar.h (hunk 1 above)
Comment (marlinstrub) [reply src/ompl/geometric/planners/bitstar/BITstar.h:485]: I agree, this is confusing. I've changed the names and added more documentation in the ABIT* class description. I also made sure ABIT*'s parameters are only declared through `declareParam` when ABIT* is instantiated (is this what you meant?).

File: src/ompl/geometric/planners/bitstar/datastructures/CostHelper.h (hunk 2 above)
Comment (marlinstrub) [reply src/ompl/geometric/planners/bitstar/datastructures/CostHelper.h]: It definitely should, nice catch!

**Comment by marlinstrub** [comment]
@mamoll Thanks for the feedback! And sorry for the slow response, I'm pushing to create some decent videos for ICRA.

I've pushed the requested changes except for adding BIT*/ABIT* to the planner unit test. I currently get a segfault on BIT* (old and new implementation) when adding it to the test. Fixing this is at the top of my list now, you can expect to hear back from me soon.

May I ask when you plan on getting the new OMPL release out? The reason I ask is that I present a second planner at this year's ICRA and I would really like to merge it upstream as well.

**Comment by mamoll** [comment]
> May I ask when you plan on getting the new OMPL release out? The reason I ask is that I present a second planner at this year's ICRA and I would really like to merge it upstream as well.

I am planning to release this in a couple of weeks. I need to time to review, test, and prepare a release, which means there isn't much time to add yet another planner. Can you create a PR within a week?

**Comment by marlinstrub** [comment]
Yes, I should be able to create the PR within the week. The implementation is done and cleaned up, I just want to do a final review myself before I push.

**Comment by marlinstrub** [comment]
Here is why there was a segfault in the unit tests of BIT* (and ABIT*): The unit test calls a planner's `setup` method after it sets the problem definition in the planner but before it fully sets up the problem (via `setupProblem`). BIT* calls `PlannerInputStates::nextGoal` in its `setup` method, which results in a segfault because `PlannerInputStates::nextGoal` implicitly assumes `pdef_->getGoal() != nullptr`.

I fixed it by making this assumption explicit, i.e., now `PlannerInputStates::nextGoal` returns `nullptr` if the problem definition does not have a valid pointer to a goal (instead of segfaulting). This is in the `Planner` base class, but I don't see how it could possibly have a negative effect on any planner. If you want, I could instead implement the check in BIT*, but then I'd suggest that we'd at least have an `assert(pdef_->getGoal() != nullptr)` in `PlannerInputStates::nextGoal`.

**Comment by marlinstrub** [comment]
There seems to be a problem with ABIT* in the `test_2dcircles_opt_geometric` that the AppVeyor CI catches. The same error occurred on my machine before c6edf74, but with this fix I can't reproduce it on my local machine. Has a similar error occurred for any other planner on AppVeyor? I'd like to fix it, but I'm not sure where to start.

**Comment by gammell** [comment]
https://ci.appveyor.com/project/mamoll/ompl/builds/32917721#L2848
> Testing ABITstar ...
> unknown location(0): fatal error: in "MyPlanTestFixture/geometric_ABITstar": memory access violation occurred at address 0x00000008, while attempting to  read inaccessible data
> C:\projects\ompl\tests\geometric\2d\2dcircles_optimize.cpp(172): last checkpoint
> *** 1 failure is detected in the test module "GeometricPlanningOpt"

Hm... Maybe try running `valgrind` on the `2dcircles_optimize.cpp` and see if it turns up anything on your machine?

**Comment by marlinstrub** [comment]
Will do!

**Comment by marlinstrub** [comment]
The test doesn't produce deterministic results on my machine (despite `RNG::setSeed` being called), but 3d67511 fixes a bug and it seems that was what was causing the issue.

**Pull request merged**